integritybackend/actions.py

```diff
@@ -47,103 +48,116 @@ def archive(self, asset_fullpath: str, org_config: dict, collection_id: str):
+        if f"{content_sha_unverified}-meta-recorder.json" not in zip_listing:
+            raise Exception(f"ZIP at {zip_path} has no recorder metadata file: {zip_listing}")
 
         # Extract content file
-        bundle_sha = os.path.splitext(content_filename)[0]
-        tmp_dir = asset_helper.get_tmp_collection_dir(collection_id, "archive")
-        zip_dir = os.path.join(tmp_dir, bundle_sha)
+        tmp_dir = asset_helper.get_tmp_action_dir(collection_id, "archive")
+        zip_dir = os.path.join(tmp_dir, content_sha_unverified)
         extracted_content = os.path.join(zip_dir, content_filename)
         file_util.create_dir(zip_dir)
         zip_util.extract_file(tmp_zip, content_filename, extracted_content)
 
-        # Generate content hashes
+        # Generate content hashes and verify
         content_sha = file_util.digest_sha256(extracted_content)
+        if content_sha != content_sha_unverified:
+            raise Exception(f"SHA-256 of content does not match file name: {zip_path}")
         content_cid = file_util.digest_cidv1(extracted_content)
         content_md5 = file_util.digest_md5(extracted_content)
-
-        # Rename to export name: SHA-256 of content file
-        final_zip = os.path.join(archive_dir, content_sha + ".zip")
-        os.rename(tmp_zip, final_zip)
+        _logger.info("Content verified for archival: {zip_path}")
 
         # Register on OpenTimestamp and add that file to zip
-        content_ots = extracted_content + ".ots"
-        file_util.register_timestamp(extracted_content, content_ots)
-        zip_util.append(
-            final_zip,
-            content_ots,
-            "proofs/" + os.path.basename(content_ots),
-        )
-
-        # Get final ZIP hashes
-        zip_sha = file_util.digest_sha256(final_zip)
-        zip_md5 = file_util.digest_md5(final_zip)
-        zip_cid = file_util.digest_cidv1(final_zip)
-
-        # Encrypt ZIP, and get those hashes
-        aes_key = crypto_util.get_key(action_config["encryption"]["key"])
-        enc_zip = os.path.join(archive_dir, content_sha + ".encrypted")
-        file_util.encrypt(aes_key, final_zip, enc_zip)
-
-        enc_zip_sha = file_util.digest_sha256(enc_zip)
-        enc_zip_md5 = file_util.digest_md5(enc_zip)
-        enc_zip_cid = file_util.digest_cidv1(enc_zip)
-
-        # TODO: step 7 and 8
-
+        if action_params["registration_policies"]["opentimestamps"]["active"]:
+            content_ots = extracted_content + ".ots"
+            file_util.register_timestamp(extracted_content, content_ots)
+            zip_util.append(
+                tmp_zip,
+                content_ots,
+                "proofs/" + os.path.basename(content_ots),
+            )
+            _logger.info(f"Content registered on opentimestamps: {content_ots}")
+        else:
+            _logger.info(f"Content registration on opentimestamps skipped")
+
+        # Get archive ZIP hashes
```

Comment:
@makeworld-the-better-one the ZIP is named `sha(ZIP)` not `sha(asset)`.

Reply:
From what I [wrote down](https://github.com/starlinglab/integrity-backend/issues/67#issuecomment-1082406996) from our previous meeting I thought it was the other way, but either works.

Reply:
>Name of final ZIP (unencrypted and encrypted) is the hash of the asset / content file

Oh you're right. I may be wrong on this. Let me fix either the docs or the implementation here and get back to you.

Reply:
@YurkoWasHere brought up a good point that we'd have collision if two same assets have different metadata in a zip somehow. I think this is an important consideration.